src/librustc/ty/sty.rs

```diff
@@ -1848,6 +1848,78 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
             _ => bug!("cannot convert type `{:?}` to a closure kind", self),
         }
     }
+
+    //
+    pub fn has_specified_layout(&self, tcx: TyCtxt<'a, 'gcx, 'tcx>) -> Option<bool> {
+        match self.sty {
+            // Not concrete, cannot say anything
+            TyInfer(_) | TyError | TyParam(_) | TyProjection(_) | TyAnon(_, _) | TyDynamic(_, _) =>
+                None,
```

Comment:
`TyDynamic` is a `dyn Trait` (trait object), so it should be treated like `TyStr` IMO.